Choice rule payload validation

kbrock opened the pull request:
alt to #187

Depend upon:

- [x] #229
-  ~#246~
- [x] #253
- [x] #254
- [x] #257
- [x] #264 
- [x] #266 
- [x] #267 
- [x] #273 
- [x] #274 
- [x] #276 
- [x] #295


Overview
========

Remove un-necessary string allocations. and symbol lookups

Most of the changes have already been pulled out and merged.
So if we don't like this, then that is fine.

---

Leaving this for legacy reasons


Overview
========

Choice does not quite behave the same as the AWS States Language reference implementation.

If you specify a `Path` that points to nothing or the wrong type, aws raises a `States.Runtime`. In truth, the type checking is not the most consistent, but the missing path is always raised.

Changes:

- Validate `Next` points to a valid state.
- Validate `Variable` and compare key values are the correct data type.
- Raise exceptions when `Variable` or compare key path values not not found or the wrong data type. It used to throw ruby exceptions.
- Fix `"IsPresent": true` to detect values present rather than not null.
- All `"Is*"` comparisons now support `true` and `false` values.
- Support `Choice` with no `Default` provided.

kbrock commented:
update:
- cop: returning `false` instead of `nil` for bad lhv
- cop: space in spec
- rebased

kbrock commented:
wip: pre-compiling operations

kbrock commented:
update:
- compiled expression into command tree
- changed key comparison to not look for String, but rather (String)(GreaterThan)(Path). It has a few false positives, (IntegerMatchesPath) but they will all probably work.

This is a lot more involved but a lot more strict/stringent

kbrock commented:
Either command line rubocop complains about having a freeze or web version complains about it missing.

un-wip: this seems to work

kbrock commented:
update:
- rebase
- simplified down choice class lookup code

update:
- added `freeze` back in

Fryguy commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -4,88 +4,36 @@ module Floe
+          "IsString"          => Floe::Workflow::ChoiceRule::IsString,
+          "IsBoolean"         => Floe::Workflow::ChoiceRule::IsBoolean,
+          "IsTimestamp"       => Floe::Workflow::ChoiceRule::IsTimestamp,
+          "Equals"            => Floe::Workflow::ChoiceRule::Data::Equals,
+          "LessThan"          => Floe::Workflow::ChoiceRule::Data::LessThan,
+          "GreaterThan"       => Floe::Workflow::ChoiceRule::Data::GreaterThan,
+          "LessThanEquals"    => Floe::Workflow::ChoiceRule::Data::LessThanEquals,
+          "GreaterThanEquals" => Floe::Workflow::ChoiceRule::Data::GreaterThanEquals,
+          "Matches"           => Floe::Workflow::ChoiceRule::Data::Matches
```
Is this right?  I don't see a Data namespace.

kbrock replied:
what the heck?
why are these tests running?

kbrock replied:
looks like it was finding `Floe::Workflow::ChoiceRule::Matches` when referencing `Floe::Workflow::ChoiceRule::Data::Matches`

Fryguy commented:
Can you add specs around these new classes?  wondering if would catch the invalid namespace thing I found.

kbrock commented:
update:
- change class reference

I commented out every line there, and the specs fail. So every one of those classes are tested

kbrock commented:
Discussion:
Instead of using a command pattern, we can change this to store the operation so we wouldn't need all these classes.

e.g.:
```
OPERATIONS = { "StringEquals" => "==" }
lhs.send(OPERATIONS[choice_value], rhs)
```

Fryguy commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -4,88 +4,36 @@ module Floe
+
+        def self.klass_params(payload)
+          payload.each_key do |key|
+            values = COMPARE_RULE.match(key)
+            return [key, DATA_RULES[values[2]], !!values[3]] if values
```
I think this would read better if the match was decomposed or perhaps if named captures were used instead of using values[2], values[3], etc.

Some other quetsions
- where is the prefix used (the `(String|Numeric|Boolean|Timestamp)` part)?  It seems to be ignored here.
- The caller does `compare_key, klass = klass_params(payload)`, but this method returns 3 items, so is that a bug?  Where did the 3rd param go with the path, and why isn't it being used?

kbrock replied:
the current pattern for the states/command objects to parse the payload themselves.
Here, we need to parse the string to 

The old code ignored the first part, so I continued to ignore it. I can see doing type checks or conversions.

returning 2 vs 3 is probably a bug. thanks

kbrock replied:
1. yes, we ignore the type prefix part and don't currently do data type conversion. (not sure if that is needed)
2. changed this code. think all set.

Fryguy commented:
> Discussion: Instead of using a command pattern, we can change this to store the operation so we wouldn't need all these classes.
> 
> e.g.:
> 
> ```
> OPERATIONS = { "StringEquals" => "==" }
> lhs.send(OPERATIONS[choice_value], rhs)
> ```

Maybe, but this all feels overly complicated to me personally.  I don't see why each of these can't be a simple method (or even just the original code the way it was).  Just curious what problem you were trying to solve.  I could see creating the classes to encapsulate a `true?` + `valid?` pair, but all of the new classes just have a single method.

kbrock commented:
update:
- rebase
- dropped comparison classes and put into an operation's hash.
- No longer modify parent `ChoiceRule` class
- storing path or value in `ref` to simplify the `initialize` code

kbrock commented:
update
- went back to all the is_{type} methods
- dropped all separate classes
- more pedantic at build time (fewer invalid cases go through

kbrock commented:
update:
- tried to reduce complexity of `initialize` but it remained the same -- 23.

kbrock commented:
update:
- another attempt at simplifying `initialize` - down to 10. think this is the best I can do

kbrock commented:
please go through commit by commit. much easier to read
update:
- validate Next has good value -- much of the complexity here
- validate Next is present for all rules (but no child rules)
- validates {String}GreaterThan uses string values in runtime (do not implicit type cast)
- more stringent compare_key validation
- broke into 3 types of comparisons, Is, StringMatch, operations

kbrock commented:
update:
- cops (spacing, hashrocket) - leaving nested `if` alone

kbrock commented:
update:
- dropped the `Next` checking. This passed around a bunch of workspaces and payloads and if desired, can be added at a later date.

kbrock commented:
update:
- rebase

kbrock commented:
update:

- removed `operation()`. now: `eq?`, `gt?`

update:

- removed `is?(value, type)` and case statement. now: `send(type_check, value)`
- removed `case operation`. now: `send(operation, lhs, rhs)`
- no longer look up left hand side if we are using unary case

kbrock commented:
update:

- testing BooleanEquals. was concerned false case would mess something up

kbrock commented:
update:
- rubocop: fixed comment on compare_value method

kbrock commented:
update:
- depend upon common validation rules
- added validation of Next (top level only, not bottom level)
- fixed some test

kbrock commented on lib/floe/workflow/choice_rule.rb:
```diff
@@ -4,31 +4,37 @@ module Floe
 
-        def build_children(sub_payloads)
-          sub_payloads.map { |payload| build(payload) }
+        def build_children(name, sub_payloads)
+          # passing workflow = nil because it is not at the top level
+          sub_payloads.map { |payload| build(nil, name, payload) }
```
🍅 Not sure that I like not passing `workflow` meaning it is a lower level. (I don't like passing name/workflow into here at all, but I digress)

Do you like `:top => false` instead?

kbrock commented:
update:
- rebase (onto newer master and #209 )
- reordered the commits. now, only the last one adds the `send(operator, lhs, rhs)`

kbrock commented:
update:
- fix cops

kbrock commented:
WIP: ran into an issue with non matching values

```
"error": "States.Runtime",
"cause": "An error occurred while executing the state 'Verify3' (entered at the event id #18). Invalid path '$.passed_in': The choice state's condition path references an invalid value."
```

kbrock commented:
WIP status:
- when paths are not valid (the parameter is missing or the path points to a different type than expected) has unpredictable runtime issues. I am writing tests to understand it better.
- Lets nail down validation framework before merging this (or remove validations?)

kbrock commented:
update:
- rebased
- fixed missing value response (I had thought it was ok, but state language ref impl clearly throws errors)
- extracted helpers for runtime errors

kbrock commented:
update:
- de-flogging
- pulled setting compare_value into own method (again, for "reducing complexity")

kbrock commented:
un-wip: this is close to the way aws state machine implementation is working.
It seems inconsistent when it throws an error. I went with a simpler / more pedantic route.

From here I would like to wrap Path(s), InputPath, and OutputPath everywhere so we have easier debugging.

I'm not displaying the values that caused the issue. we can decide if we want that to leak.

kbrock commented:
I was able to simplify validate_rule - but it was too hard to read it, so I went back to the simpler but more "complex" version.

I also fixed the operator version of this (it is one additional commit). Not sure how long this will live but...
[`send` method version](https://github.com/kbrock/floe/compare/choice_rule_payload_validation..choice_rule_payload_validation-send)

kbrock commented on spec/workflow/intrinsic_function_spec.rb:
```diff
@@ -103,9 +103,10 @@
+      # TODO: test this
       it "handles invalid path references" do
-        result = described_class.value("States.Array($.xxx)", {"context" => {"baz" => "qux"}}, {"input" => {"foo" => "bar"}})
-        expect(result).to eq([nil])
+        ctx = {"context" => {"baz" => "qux"}}, {"input" => {"foo" => "bar"}}
+        expect { described_class.value("States.Array($.xxx)", ctx) }.to raise_error(Floe::PathError, "references an invalid value")
```
ok, verified this:

<details>
<summary>asl file with tests</summary>

```asl
{
  "Comment": "Testing Pass intrinsics - sample input: {test: 1}",
  "StartAt": "Setup",
  "States": {
    "Setup": {
      "Comment": "(Implied test) Result => output:Result",
      "Type": "Pass",
      "Next": "ChooseTest",
      "Parameters": {"test.$": "$.test", "num1": 1, "str1": "1", "num2": 1, "null": null, "str2": "1"}
    },
    "ChooseTest": {
      "Type":  "Choice",
      "Choices": [
        {"Variable": "$.test", "NumericEquals": 1, "Next": "Test1"},
        {"Variable": "$.test", "NumericEquals": 2, "Next": "Test2"}
      ],
      "Default": "Default"
    },
    "Test1": {
      "Comment": "Variable missing -> The function 'States.Array($.missing)' had the following error: The JsonPath argument for the field '$.missing' could not be found in the input ...",
      "Type": "Pass",
      "Parameters": {
        "param1.$": "States.Array($.missing)"
      },
      "Next": "Match"
    },
    "Test2": {
      "Comment": "output: {param1: [null]}",
      "Type": "Pass",
      "Parameters": {
        "param1.$": "States.Array($.null)"
      },
      "Next": "Verify2"
    },
    "Verify2": {
      "Type": "Choice",
      "Choices": [
        {
          "Or": [
            {"Variable": "$.param1[0]", "IsPresent": true},
            {"Variable": "$.param1[0]", "IsNull": true},
            {"Variable": "$.param1[1]", "IsPresent": false}
          ], "Next": "Match"
        }
      ],
      "Default": "Default"
    },
    "Match": {
      "Type": "Pass",
      "End": true
    },
    "Default": {
      "Type": "Pass",
      "End": true,
      "Result": "Nothing matched"
    }
  }
}
```

</details>

Fryguy replied:
yeah that fails in the simulator - can you verify what a path does without the intrinsic function but with a missing key?  If that also fails then I think we should fail this in the Path type.

agrare commented on lib/floe/workflow/path.rb:
```diff
@@ -34,7 +34,18 @@ def value(context, input = {})
-        results.count < 2 ? results.first : results
+        case results.count
+        when 0
+          raise Floe::PathError, "references an invalid value"
```
@kbrock does this match what StepFunctions does?  Or would this just return `null`?

kbrock replied:
Both Intrinsics and Choice threw errors if the path is not present (except for IsPresence checks).
That is where I found the issue with `IsPresent` vs `IsNull` - present checks that it is defined vs null checks that it is defined, but null.

Now datatype comparison was inconsistent, but presence was consistent.

agrare commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -40,52 +61,91 @@ def true?(context, input)
+
+          # path found the variable_value, (so if they said true, return true)
+          ret_value
+        rescue Floe::ExecutionError
```
You're rescuing ExecutionError here but in raise_runtime_error! https://github.com/ManageIQ/floe/pull/189/files#diff-d68e675c8b56044da38fd89268ac291a59a7475ae4c31d98a09e76e041092fe3R27 you raise a PathError is that right?

agrare replied:
Oh nevermind it looks like you catch PathError then re-raise ExceutionError

kbrock replied:
not sure a better way to deal with the whole path errors. I want good error messaging saying where the path was defined and the problem. For both runtime and parse time issues.

agrare commented on lib/floe/workflow/choice_rule.rb:
```diff
@@ -42,8 +44,18 @@ def true?(*)
+
+      # returns true if this is a child rule underneath an And/Or/Not
+      def is_child? # rubocop:disable Naming/PredicateName
+        %w[And Or Not].include?(name[-3])
```
This looks a little fragile, maybe check array union `(name & %w[And Or Not]).empty?`

kbrock replied:
ok, added a spec and this works:
```ruby
      def is_child? # rubocop:disable Naming/PredicateName
        !(%w[And Or Not] & name[0..-2]).empty?
      end
```

Kind of prefer `name[-3]`, though it takes some thinking. This other way requires less thinking. Which is always prefered

agrare replied:
Yeah thinking about it some more I like that name[-3] is more exact but I definitely did a triple take reading it. Maybe a comment about what the name array looks like so it is more obvious

agrare commented on lib/floe/workflow/states/choice.rb:
```diff
@@ -21,11 +21,19 @@ def initialize(workflow, name, payload)
+          # input / output paths were bad
+          # choice state was bad
+          context.next_state = nil
+          context.output     = {"Error" => "States.RuntimeError", "Cause" => e.message}
```
Would this be a good time to use `States.NoChoiceMatched`?
> A Choice State failed to find a match for the condition field extracted from its input.

kbrock replied:
this caused a little more change than I wanted - but great idea.

kbrock commented:
update:
- reworked validation mixin to get error messages out (and reused in "Cause")
- changed `ChoiceRule#is_child?`
- added States.NoChoiceMatched (and kept RuntimeError for when the paths are bad)

agrare commented:
Looks good overall I think we do need to do the `States.NoChoiceMatched` change though since there is no `States.RuntimeError`

kbrock commented:
update:
- ruby 3.0 compliant (removed `...`)

kbrock commented:
update:
- "reduced complexity"

kbrock commented:
update:
- put back in is_present?, thought it is a no-op
- fix duplicate `variable_value` (rebase introduced this)

update
- PathError is not a type of ExecutionError - easier to catch errors

kbrock commented:
update:
- reduce complexity of Data#initialize

kbrock commented on lib/floe.rb:
```diff
@@ -44,6 +44,17 @@ class Error < StandardError; end
+  class ExecutionError < Error
+    attr_reader :floe_error
+
+    def initialize(message, floe_error = "States.RuntimeError")
```
Thanks. per https://docs.aws.amazon.com/step-functions/latest/dg/concepts-error-handling.html#states-runtime-error that should be `States.Runtime`

There seem to be quite a few trapable errors

agrare replied:
Ah thanks that's a good list I'll bookmark that.  Should this be `States.Runtime` then not `States.RuntimeError`?

kbrock replied:
Thanks for finding that. I've pushed a new version with that corrected
Converting from (ruby) RuntimeError to States.Runtime got me flubbed in a few places.

kbrock commented:
update:
- rebased

kbrock commented:
update:
- rebase
- split 1 commit into 2 (moving @variable from choice_rule to data in 2 steps now vs 1)

kbrock commented:
update:
- fix complexity of `State#run_nonblock` by introducing `mark_exception`

(that hopefully will be the last code climate issue with this code)

kbrock commented:
update:
- rebase
- renamed UNARY_RULE to TYPE_CHECK
- renamed BINARY_RULE to OPERATION
- added some comments around `@type` and `@path` variables

just superficial changes. Code hasn't changed

kbrock commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -4,19 +4,35 @@ module Floe
           rhs = compare_value(context, input)
 
-          validate!(lhs)
-
```
Before:
- fetches a value, and verifies what ever came out is not null

After:
- fetches a value and verifies it existed in the input hash

kbrock commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -4,19 +4,35 @@ module Floe
-          when "IsTimestamp" then is_timestamp?(lhs)
+          when "IsNull" then is_null?(lhs, rhs)
+          when "IsNumeric" then is_numeric?(lhs, rhs)
+          when "IsString" then is_string?(lhs, rhs)
```
Before: `"IsString" => "*"` verified it is a string.
After: `"IsString" => true` works that way, but `"IsString" => false` does the opposite
(across all checks)

kbrock commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -40,52 +56,104 @@ def true?(context, input)
 
-        def compare_key
-          @compare_key ||= payload.keys.detect { |key| key.match?(/^(IsNull|IsPresent|IsNumeric|IsString|IsBoolean|IsTimestamp|String|Numeric|Boolean|Timestamp)/) }
+        def variable_value(context, input)
+          fetch_path("Variable", variable, context, input)
         end
```
(moved over from ChoiceRule since this is data specific)
Before:
- convert `@variable` to a `Path` at runtime and fetch value.
After:
- `@variable is a `Path`, so bad Path detected at initialization time
- Path not present in input is a runtime error
- Data type of input is checked.

kbrock commented on lib/floe/workflow/path.rb:
```diff
@@ -34,7 +34,18 @@ def value(context, input = {})
-        results.count < 2 ? results.first : results
+        case results.count
+        when 0
+          raise Floe::PathError, "Path [#{payload}] references an invalid value"
```
This one change has big consequences.

Before: A path pointing to nothing would just treat the value as `nil`
After: A path pointing to nothing is a runtime error

kbrock commented on lib/floe/workflow/state.rb:
```diff
@@ -48,17 +48,27 @@ def run_nonblock!(context)
         return Errno::EAGAIN unless ready?(context)
 
         finish(context)
+      rescue Floe::ExecutionError => e
+        mark_error(context, e)
```
This was detected because tests running against a state would work, but running against a workflow would not.

Before:
- Errors needed to be caught by each `State`
- Errors resulted in a state that was not complete.
- Errors did not always get their way into output.

After:
- runtime errors always set the context accordingly.

kbrock commented:
WIP: piece by piece pulling out PRs from this

kbrock commented:
update:
- fix guard clause - ensured bad operators found in initializer not runtime

update:
- fix `}` spacing cop

kbrock commented:
update:
- rebased

update:
- converted `"string".to_sym` to `:"string"`

Fryguy commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -6,12 +6,14 @@ class ChoiceRule
       class Data < Floe::Workflow::ChoiceRule
         TYPES      = ["String", "Numeric", "Boolean", "Timestamp", "Present", "Null"].freeze
         COMPARES   = ["Equals", "LessThan", "GreaterThan", "LessThanEquals", "GreaterThanEquals", "Matches"].freeze
+        OPERATIONS = TYPES.each_with_object({}) { |dt, a| a[dt] = :"is_#{dt.downcase}?" } \
```
Line continuation is not needed

```suggestion
        OPERATIONS = TYPES.each_with_object({}) { |dt, a| a[dt] = :"is_#{dt.downcase}?" }
```

Fryguy commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -6,12 +6,14 @@ class ChoiceRule
         TYPE_CHECK = /^(Is)(#{TYPES.join("|")})$/
         # e.g.: (String)(LessThan)(Path), (Numeric)(GreaterThanEquals)()
         OPERATION  = /^(#{(TYPES - %w[Null Present]).join("|")})(#{COMPARES.join("|")})(Path)?$/
 
-        attr_reader :variable, :compare_key, :operation, :type, :compare_predicate, :path
+        attr_reader :variable, :compare_key, :operator, :type, :compare_predicate, :path
```
This is changing the "public" api - not a big deal if it's internal, but is anything else using this?

Fryguy replied:
No it's fine, I just wanted to make sure that was accounted for or not.

Fryguy commented:
This PR got renamed? Having trouble following the history here.

kbrock commented:
This was a PR with a ton of stuff. almost all of the stuff had been extracted (as you can see from the punch list at the top)
The only thing left in a minimal (and not mandatory) change to not build the names of the operations at runtime, but rather just use a lookup.

Originally, the type checks (`IsString`) and operations/compares (`StringGreaterThan`) stored in `TYPES` and `COMPARES` were hashes from the keys to the method names. As we extracted out the PRs, this hash structure was lost.

I am not a fan of the non-grep style of `send("#{var1}#{var2}")`. I guess `OPERATIONS` is essentially just that, thought defined up front. Alternately, explicitly listing the operations in the hash form of `TYPES` and `COMPARES` may make sense.


update:
- rebased
- spelling fixes in comments
- some descriptions around the checks and parameters
- removed unnecessary line continuation
-

Fryguy commented on lib/floe/workflow/choice_rule/data.rb:
```diff
@@ -110,22 +112,20 @@ def parse_compare_key
             elsif (match_value = TYPE_CHECK.match(key))
               @compare_key = key
-              _operator, type = match_value.captures
+              _is, @operator = match_value.captures
```
If we don't need the `Is`, you can also just not capture it in the first place
